python/pyarrow/_substrait.pyx

```diff
@@ -102,3 +103,40 @@ def get_supported_functions():
     for c_id in c_ids:
         functions_list.append(frombytes(c_id))
     return functions_list
+
+
+cdef shared_ptr[CTable] _process_named_table(dict named_args, const std_vector[c_string]& names):
+    cdef c_string c_name
+    py_names = []
+    for i in range(names.size()):
+        c_name = names[i]
+        py_names.append(frombytes(c_name))
+    return pyarrow_unwrap_table(named_args["provider"](py_names))
+
+
+def run_query_with_provider(plan, table_provider):
```

Comment:
Can we make `table_provider` an optional argument for `run_query` rather than adding a new method?  I don't know if we can do it in cython (I think we can for a `def` method) or not but ideally the python API wouldn't need two methods.

Reply:
Of course, this was just a prototype code to show case for test purpose in a short time. I already refactored the code. I think it covers what you are asking for.